File: cpp/deinflector.cpp
#include "deinflector.h"
#include "word_types.h"

#include <list>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <cassert>


Deinflector::Deinflector(const string_view& deinflect, const string_view& expressions)
{
	const char* p = deinflect.data;
	const char* end = deinflect.data + deinflect.length;
	while (p < end)
	{
		if (*p == '#' || *p == ' ' || *p == '\t' || *p == '\n')
		{
			p = std::find(p, end, '\n') + 1;
			continue;
		}
		const char* const line_end = std::find(p, end, '\n');

		std::string from;
		Rule rule;
		int column = 0;
		while (p < line_end)
		{
			const char* const column_end = std::find(p, line_end, '\t');
			column += 1;
			switch (column)
			{
				case 1:
					from = std::string(p, column_end);
					break;
				case 2:
					rule.to = std::string(p, column_end);
					break;
				case 3:
					rule.source_type = inflection_type_to_int(p, column_end);
					break;
				case 4:
					rule.target_type = inflection_type_to_int(p, column_end);
					break;
				case 5:
					rule.reason = std::string(p, column_end);
					break;
				default:
					std::cerr << "Invalid deinflect rule: " << std::string(p, line_end) << std::endl;
					break;
			}
			p = column_end + 1;
		}
		while (from.length() >= rules_groups.size())
		{
			rules_groups.push_back(SuffixLengthRulesGroup(rules_groups.size()));
		}
		rules_groups.at(from.length()).rules.insert(std::make_pair(from, rule));

		p = line_end + 1;
	}

	uint32_t all_verbs_type = 0;
	for (auto& kv : INFLECTION_TYPES)
	{
		if (kv.first[0] == 'v')
		{
			all_verbs_type |= kv.second;
		}
	}
	INFLECTION_TYPES["v"] = all_verbs_type;

	std::vector<std::string> from;
	Rule rule(true);
	p = expressions.data;
	end = expressions.data + expressions.length;
	while (p < end)
	{
		if (*p == '#' || *p == ' ' || *p == '\t' || *p == '\n')
		{
			p = std::find(p, end, '\n') + 1;
			continue;
		}
		const char* const line_end = std::find(p, end, '\n');

		int column = 0;
		while (p < line_end)
		{
			const char* const column_end = std::find(p, line_end, '\t');
			column += 1;
			switch (column)
			{
				case 1:
					assert(p < column_end);
					from.push_back(std::string(p, column_end));
					break;
				case 2:
					rule.source_type = inflection_type_to_int(p, column_end);
					break;
				case 3:
					rule.after_type = *p == '*'? ANY_TYPE
					                           : inflection_type_to_int(p, column_end);
					break;
				case 4:
					while (p < column_end)
					{
						const char* const part_end = std::find(p, column_end, '|');
						assert(p < part_end);
						rule.after_form.insert(std::string(p, part_end));
						p = part_end + 1;
					}
					break;
				case 5:
					assert(p < column_end);
					if (std::all_of(p, column_end, isdigit))
					{
						rule.offset = uint32_t(std::stoul(std::string(p, column_end)));
					}
					break;
				case 6:
					while (p < column_end)
					{
						const char* const part_end = std::find(p, column_end, '|');
						const char* coma = std::find(p, part_end, ',');

						assert(p < coma);
						assert(coma + 1 < part_end);
						rule.sense_indices.push_back(std::make_pair(
							std::stoi(std::string(p, coma)),
							std::stoi(std::string(coma + 1, part_end))
						));
						p = part_end + 1;
					}
					break;
				default:
					std::cerr << "Invalid expression: " << std::string(p, line_end) << std::endl;
					break;
			}
			p = column_end + 1;
		}
		if (column == 1)
		{
			p = line_end + 1;
			continue;
		}

		for (auto& from_string : from)
		{
			while (from_string.length() >= rules_groups.size())
			{
				rules_groups.push_back(SuffixLengthRulesGroup(rules_groups.size()));
			}
			rules_groups.at(from_string.length()).rules.insert(std::make_pair(from_string, rule));
		}
		rule = Rule(true);
		from.clear();

		p = line_end + 1;
	}

	rules_groups.erase(std::remove_if(rules_groups.begin(), rules_groups.end(), [](SuffixLengthRulesGroup& r) {
		return r.rules.empty();
	}), rules_groups.end());
}

std::list<Candidate> Deinflector::deinflect(const std::string& word)
{
	PROFILE
//	std::cout << "deinflect( " << word << " )" << std::endl;

	std::list<Candidate> r;
	std::map<std::string, Candidate*> have;
	r.emplace_back(word,
	// Original word can have any type
		ANY_TYPE);
	have[word] = &*r.begin();


	for (auto candidate_it = r.begin(); candidate_it != r.end(); ++candidate_it)
	{
		const Candidate& candidate = *candidate_it;

		for (size_t j = 0; j < rules_groups.size(); ++j)
		{
			const SuffixLengthRulesGroup& g = rules_groups.at(j);
			if (g.suffix_length > candidate.word.length()) break;

			const std::string end = candidate.word.substr(candidate.word.length() - g.suffix_length);
			auto range = g.rules.equal_range(end);
			if (range.first == g.rules.end()) continue;

			for (auto it = range.first; it != range.second; ++it)
			{
				const Rule& rule = it->second;
				// If rule isn't applicable to this word
				if ((candidate.type & rule.source_type) == 0) continue;
				if (!candidate.expected_forms.empty()
						&& candidate.expected_forms.count(rule.reason) == 0
						&& candidate.expected_forms.count("*") == 0)
				{
					continue;
				}

				std::string new_word = candidate.word.substr(0, candidate.word.length() - g.suffix_length);
				if (!rule.expression)
				{
					// Inflection
					new_word += rule.to;
					if (new_word.length() <= 1) continue;
					auto have_it = have.find(new_word);
					if (have_it != have.end())
					{
						// If we have already obtained this new word through other
						// chain of deinflections - update only it's type.
						// We union types to denote that it can have any of these types
						have_it->second->type |= rule.target_type;
						continue;
					}

					r.emplace_back(new_word, rule.target_type);
					if (candidate.expected_forms.count(rule.reason) == 0)
					{
						if (candidate.reason.length() > 0)
						{
							r.back().reason = rule.reason + " &lt; " + candidate.reason;
						}
						else
						{
							r.back().reason = rule.reason;
						}
					}
					r.back().expressions = candidate.expressions;
					r.back().expressions_forms = candidate.expressions_forms;
					have[new_word] = &r.back();

				}
				else if (g.suffix_length < candidate.word.length())
				{

					// Expression
					if (!candidate.expressions.empty() &&
							(candidate.expressions.back()->after_type & rule.source_type) == 0)
					{
						continue;
					}

					size_t numSpecials = 0;
					if (rule.after_form.count("negative stem") > 0)
					{
						numSpecials += 1;
						auto new_special = new_word + u8"ない";
						if (new_word == u8"せ")
						{
							new_special = u8"しない";
						}
						if (have.count(new_special) > 0)
						{
							std::cerr << "Well, that's strange: " << word << " " << new_special << std::endl;
							continue;
						}

						r.emplace_back(new_special, INFLECTION_TYPES["adj-i"],
							candidate.expressions_forms, end,
							candidate.expressions, &rule,
							std::set<std::string>{"negative", "masu stem"}
						);
						have[new_special] = &r.back();
					}

					if (rule.after_form.count("provisional stem") > 0)
					{
						numSpecials += 1;
						auto new_special = new_word + u8"ば";
						if (have.count(new_special) > 0)
						{
							std::cerr << "Well, that's strange: " << word << " " << new_special << std::endl;
							continue;
						}

						r.emplace_back(new_special, INFLECTION_TYPES["raw"],
							candidate.expressions_forms, end,
							candidate.expressions, &rule,
							std::set<std::string>{"-ba"}
						);
						have[new_special] = &r.back();
					}

					if (rule.after_form.count("adjective stem") > 0)
					{
						numSpecials += 1;
						auto new_special = new_word + u8"い";
						if (have.count(new_special) > 0)
						{
							std::cerr << "Well, that's strange: " << word << " " << new_special << std::endl;
							continue;
						}

						r.emplace_back(new_special, INFLECTION_TYPES["adj-i"],
							candidate.expressions_forms, end,
							candidate.expressions, &rule,
							std::set<std::string>{"negative"}
						);
						have[new_special] = &r.back();
					}

					if (numSpecials < rule.after_form.size())
					{
						if (have.count(new_word) > 0)
						{
							std::cerr << "Well, that's strange: " << word << " " << new_word << std::endl;
							continue;
						}

						r.emplace_back(new_word, ANY_TYPE,
							candidate.expressions_forms, end,
							candidate.expressions, &rule,
							rule.after_form
						);
						have[new_word] = &r.back();
					}
				}
			}
		}
	}

	for (auto it = r.begin(); it != r.end();)
	{
		if (!it->expressions.empty())
		{
			if ((it->expressions.back()->after_type & it->type) == 0)
			{
				it = r.erase(it);
				continue;
			}
			it->type &= it->expressions.back()->after_type;
		}
		++it;
	}
	return r;
}

File: js/deinflector.js
"use strict";

class Deinflector {

	constructor(rulesArray, expressions) {
		const rules = [];
		for (var i = 0; i < rulesArray.length; ++i) {
			const columns = rulesArray[i].split('\t');
			if (columns.length !== 5) {
				console.error('Invalid line in `deinflect.dat`:', rulesArray[i]);
				continue;
			}

			var rulesGroupByLength = rules[columns[0].length];
			if (!rulesGroupByLength) {
				rulesGroupByLength = rules[columns[0].length] = {
					suffixLength: columns[0].length
				}
			}
			if (!(columns[0] in rulesGroupByLength)) {
				rulesGroupByLength[columns[0]] = [];
			}

			const rule = {
				from: columns[0],
				to: columns[1],
				sourceType: inflectionTypeToInt(columns[2].split('|'), true),
				targetType: inflectionTypeToInt(columns[3].split('|'), true),
				reason: columns[4]
			};
			rulesGroupByLength[rule.from].push(rule);
		}

		var allVerbsType = 0;
		for (var k in INFLECTION_TYPES) {
			if (k.startsWith('v')) {
				allVerbsType |= INFLECTION_TYPES[k];
			}
		}
		INFLECTION_TYPES['v'] = allVerbsType;

		var from = [];
		for(var i = 0; i < expressions.length; ++i) {
			const columns = expressions[i].split('\t');
			from.push(columns[0]);
			if (columns.length === 1) {
				continue;
			}
			if (columns.length != 6) {
				console.error('Invalid expression line:', expressions[i]);
				from = [];
				continue;
			}
			const rule = {
				from,
				sourceType: INFLECTION_TYPES[columns[1]],
				afterType: columns[2] === '*' ? INFLECTION_TYPES.__any__ : inflectionTypeToInt(columns[2].split('|'), true),
				//afterForms: columns[3] === '*' ? null : new Set(columns[3].split('|')),
				afterForms: new Set(columns[3].split('|')),
				offset: parseInt(columns[4]),
				senseIndices: columns[5]
			};

			for (var suffix of from) {
				var rulesGroupByLength = rules[suffix.length];
				if (!rulesGroupByLength) {
					rulesGroupByLength = rules[suffix.length] = {
						suffixLength: suffix.length
					}
				}
				if (!(suffix in rulesGroupByLength)) {
					rulesGroupByLength[suffix] = [];
				}
				rulesGroupByLength[suffix].push(rule);
			}
			from = [];
		}

		this._rules = rules.filter(o => !!o);
		//console.log(this._rules);
		this._expressionReplacementRules = [
			{
				suffix: '\u3070',
				form: 'provisional stem',
				expectedForms: new Set(['-ba'])
			},
			{
				suffix: '\u306a\u3044',
				form: 'negative stem',
				expectedForms: new Set(['negative'])
			},
			{
				suffix: '\u3044',
				form: 'adjective stem',
				type: INFLECTION_TYPES['adj-i'],
				expectedForms: new Set('negative')
			},
			{
				suffix: '',
				form: null
			}
		];
	}

	_checkExpression(conj) {
		var res = !conj.expressions || (conj.expressions[0].afterType & conj.type) !== 0;
		if (res && conj.expressions) {
			conj.expression = conj.expressions.map(e => e.from[0]).join('+');
			conj.type &= conj.expressions[0].afterType;
		}
		return res;
	}

	deinflect(word) {
		//console.log('deinflect(', word, ')');
		var r = [];
		var have = {};
		r.push({
			word,
		// Original word can have any type
			type: INFLECTION_TYPES.__any__,
			reason: ''
		});
		have[word] = 0;


		var i = -1;
		while (++i < r.length) {
			word = r[i].word;
			var wordLen = word.length;
			var type = r[i].type;

			for (var j = 0; j < this._rules.length; ++j) {
				var g = this._rules[j];
				if (g.suffixLength > wordLen) continue;

				var end = word.substr(-g.suffixLength);
				if (!(end in g)) continue;

				var rules = g[end];
				for (var k = 0; k < rules.length; ++k) {
					var rule = rules[k];
					// If rule isn't applicable to this word
					if ((type & rule.sourceType) === 0) continue;
					if (('expectedForms' in r[i])
							&& !r[i].expectedForms.has(rule.reason)
							&& !r[i].expectedForms.has('*')) {
						continue;
					}

					var newWord = word.substr(0, word.length - g.suffixLength);
					if ('to' in rule) {
						// Inflection
						newWord += rule.to;
						if (newWord.length <= 1) continue;
						if (newWord in have) {
							// If we have already obtained this new word through other
							// chain of deinflections - update only it's type.
							// We union types to denote that it can have any of these types
							r[have[newWord]].type = (o.type | rule.targetType);
							continue;
						}
						have[newWord] = r.length;

						var o = {};
						if (r[i].reason.length > 0) {
							o.reason = rule.reason + ' &lt; ' + r[i].reason;
						} else {
							o.reason = rule.reason;
						}
						if ('expressions' in r[i]) {
							o.expressions = r[i].expressions.slice();
						}
						// We know exact type of deinflected word -
						// it is specified by the rule we just applied
						o.type = rule.targetType;
						o.word = newWord;
						r.push(o);

					} else if (g.suffixLength < wordLen) {
						// Expression
						if ('expressions' in r[i] && (r[i].expressions[0].afterType & rule.sourceType) === 0) {
							continue;
						}

						var numSpecials = 0;
						if (rule.afterForms.has('negative stem')) {
							numSpecials += 1;
							var newSpecial = newWord + '\u306a\u3044';
							if (newWord === 'せ') {
								newSpecial = '\u3057\u306a\u3044';
							}
							if (newSpecial in have) {
								console.error("Well, that's strange", word, newSpecial);
								continue;
							}
							have[newSpecial] = r.length;

							r.push({
								reason: '',
								expressions: [rule],
								type: INFLECTION_TYPES['adj-i'],
								word: newSpecial,
								expectedForms: new Set(['negative', 'masu stem'])
							});
							if ('expressions' in r[i]) {
								Array.prototype.push.apply(r[r.length - 1].expressions, r[i].expressions);
							}
						}

						if (rule.afterForms.has('provisional stem')) {
							numSpecials += 1;
							const newSpecial = newWord + '\u3070';
							if (newSpecial in have) {
								console.error("Well, that's strange", word, newSpecial);
								continue;
							}
							have[newSpecial] = r.length;

							r.push({
								reason: '',
								expressions: [rule],
								type: INFLECTION_TYPES['raw'],
								word: newSpecial,
								expectedForms: new Set('-ba')
							});
							if ('expressions' in r[i]) {
								Array.prototype.push.apply(r[r.length - 1].expressions, r[i].expressions);
							}
						}

						if (rule.afterForms.has('adjective stem')) {
							numSpecials += 1;
							const newSpecial = newWord + '\u3044';
							if (newSpecial in have) {
								console.error("Well, that's strange", word, newSpecial);
								continue;
							}
							have[newSpecial] = r.length;

							r.push({
								reason: '',
								expressions: [rule],
								type: INFLECTION_TYPES['adj-i'],
								word: newSpecial,
								expectedForms: new Set('negative')
							});
							if ('expressions' in r[i]) {
								Array.prototype.push.apply(r[r.length - 1].expressions, r[i].expressions);
							}
						}

						if (numSpecials < rule.afterForms.size) {
							if (newWord in have) {
								console.error("Well, that's strange", word, newWord);
								continue;
							}
							have[newWord] = r.length;

							r.push({
								reason: '',
								expressions: [rule],
								type: INFLECTION_TYPES.__any__, //& rule.afterType,
								word: newWord,
								expectedForms: rule.afterForms
							});
							if ('expressions' in r[i]) {
								Array.prototype.push.apply(r[r.length - 1].expressions, r[i].expressions);
							}
						}
					}
				}
			}
		}

		//console.log(r);
		r =  r.filter(this._checkExpression, this);
		//console.log(r);
		return r;
	}


}
